File: 310.Minimum Height Trees/Solution.py
class Solution(object):
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        if n <= 1:
            return [0]
        degrees = [0] * n
        graph = {}
        for i in range(0,n):
            graph[i] = []
        for edge in edges:
            degrees[edge[1]] += 1
            degrees[edge[0]] += 1
            graph[edge[1]].append(edge[0])
            graph[edge[0]].append(edge[1])

        queue = []
        ret = []
        for j in range(0,n):
            if degrees[j] == 1:
                queue.append(j)
        while queue:
            temp = []
            ret = queue[:]
            for x in queue:
                for n in graph[x]:
                    degrees[n] -= 1
                    if degrees[n] == 1:
                        temp.append(n)
            queue = temp
        return ret

File: 310.Minimum Height Trees/src/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (n <= 1) {
        	List<Integer> re = new ArrayList<Integer>();
        	re.add(0);
        	return re;
        }
        int degrees[] = new int[n];
        List<List<Integer>> graph = new ArrayList<List<Integer>>();
        for (int i = 0; i < n; i++) {
        	graph.add(new ArrayList<Integer>());
        }
        for (int i = 0; i < edges.length; i++) {
        	degrees[edges[i][0]] += 1;
        	degrees[edges[i][1]] += 1;
        	graph.get(edges[i][0]).add(edges[i][1]);
        	graph.get(edges[i][1]).add(edges[i][0]);
        }
        
        List<Integer> queue = new ArrayList<Integer>();
        List<Integer> ret = new ArrayList<Integer>();
        for (int i = 0; i < n; i++) {
        	if (degrees[i] == 1)
        		queue.add(i);
        }
        while (!queue.isEmpty()) {
        	List<Integer> temp = new ArrayList<Integer>();
        	ret = queue;
        	for (int i = 0; i < queue.size(); i++) {
        		int x = queue.get(i);
        		List<Integer> edge = graph.get(x);
        		for (int j = 0; j < edge.size(); j++) {
        			degrees[edge.get(j)] -= 1;
        			if (degrees[edge.get(j)] == 1)
            			temp.add(edge.get(j));
        		}
        	}
        	queue = temp;
        }
    	return ret;
    }
}